File: ActionDraw_Event/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace ActionDraw_Event
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent(); 
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            this.vdSC.BaseControl.ActiveDocument.ShowUCSAxis = false;
            this.vdSC.BaseControl.ActiveDocument.FreezeEntityDrawEvents.Push(false);
            this.vdSC.BaseControl.ActiveDocument.OnActionDraw += new VectorDraw.Professional.vdObjects.vdDocument.ActionDrawEventHandler(ActiveDocument_OnActionDraw);
            this.vdSC.BaseControl.ActiveDocument.OnFigureDrawGrips += new VectorDraw.Professional.vdObjects.vdDocument.FigureDrawGripsEventHandler(ActiveDocument_OnFigureDrawGrips);
            this.vdSC.BaseControl.ActiveDocument.GlobalRenderProperties.GripSize = 20;
        }

        void ActiveDocument_OnFigureDrawGrips(object sender, VectorDraw.Render.vdRender render, ref bool cancel)
        {
            if (!chkBox_GripDraw.Checked) return;
            if (sender == null) return;

            VectorDraw.Geometry.Matrix mat = new VectorDraw.Geometry.Matrix(render.CurrentMatrix);
            render.PushToViewMatrix(); // you need to PopMatrix before ending this procedure
           
            try // here draw the grips as filled circles with transparency -->
            {
                VectorDraw.Professional.vdPrimaries.vdFigure fig = sender as VectorDraw.Professional.vdPrimaries.vdFigure;
                VectorDraw.Geometry.gPoints gripPTs;
                VectorDraw.Professional.vdFigures.vdCircle circ = new VectorDraw.Professional.vdFigures.vdCircle();
                circ.SetUnRegisterDocument(this.vdSC.BaseControl.ActiveDocument);
                circ.setDocumentDefaults();
                circ.PenColor = new VectorDraw.Professional.vdObjects.vdColor(Color.Red);
                circ.HatchProperties = new VectorDraw.Professional.vdObjects.vdHatchProperties(VectorDraw.Professional.Constants.VdConstFill.VdFillModeSolid);
                circ.Radius = render.PixelSize * render.GlobalProperties.GripSize / 2.0d;
                circ.HatchProperties.FillColor.SystemColor = Color.Red;
                circ.HatchProperties.FillColor.AlphaBlending = 100;
                circ.HatchProperties.Update();
                gripPTs = fig.GetGripPoints();
                VectorDraw.Geometry.gPoint tmpPt = new VectorDraw.Geometry.gPoint();
                foreach (VectorDraw.Geometry.gPoint gripPT in gripPTs)
                {
                    tmpPt = mat.Transform(gripPT);
                    circ.Center = tmpPt;
                    circ.Draw(render);
                    circ.Update();
                }
                cancel = true;
            }
            catch { }// <-- here draw the grips as filled circles with transparency

            render.PopMatrix(); // and here is the PopMatrix
        }

        void ActiveDocument_OnActionDraw(object sender, object action, bool isHideMode, ref bool cancel)
        {
            if (!chkBox.Checked) return;
            if (!(action is VectorDraw.Actions.ActionGetRefPoint)) return;
            VectorDraw.Actions.BaseAction act = action as VectorDraw.Actions.BaseAction;
            VectorDraw.Geometry.gPoint refpoint = act.ReferencePoint;
            VectorDraw.Geometry.gPoint currentpoint = act.OrthoPoint;
            VectorDraw.Professional.vdFigures.vdCircle circle = new VectorDraw.Professional.vdFigures.vdCircle();
            circle.SetUnRegisterDocument(vdSC.BaseControl.ActiveDocument);
            circle.setDocumentDefaults();
            circle.Center = VectorDraw.Geometry.gPoint.MidPoint(refpoint, currentpoint);
            circle.Radius = circle.Center.Distance3D(refpoint);
            circle.Draw(act.Render);
        }

        private void btLine_Click(object sender, EventArgs e)
        {
            this.vdSC.BaseControl.ActiveDocument.CommandAction.CmdLine("USER");
        }

        private void chkBox_GripDraw_CheckedChanged(object sender, EventArgs e)
        {
            this.vdSC.BaseControl.Redraw();
        }

        private void chkBox_CheckedChanged(object sender, EventArgs e)
        {
            this.vdSC.BaseControl.Redraw();
        }
    }
}

File: VB/ActionDraw_event_VB/Form1.vb
Public Class Form1

    WithEvents doc As VectorDraw.Professional.vdObjects.vdDocument

    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        doc = vdSC.BaseControl.ActiveDocument
        Me.vdSC.BaseControl.ActiveDocument.ShowUCSAxis = False
        Me.vdSC.BaseControl.ActiveDocument.FreezeEntityDrawEvents.Push(False)
    End Sub

    Private Sub btLine_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btLine.Click
        Me.vdSC.BaseControl.ActiveDocument.CommandAction.CmdLine("USER")
    End Sub

    Private Sub doc_OnActionDraw(ByVal sender As Object, ByVal action As Object, ByVal isHideMode As Boolean, ByRef cancel As Boolean) Handles doc.OnActionDraw
        If Not (chkBox.Checked) Then Return
        If Not (TypeOf action Is VectorDraw.Actions.ActionGetRefPoint) Then Return
        Dim act As VectorDraw.Actions.BaseAction = action
        Dim refpoint As VectorDraw.Geometry.gPoint = act.ReferencePoint
        Dim currentpoint As VectorDraw.Geometry.gPoint = act.OrthoPoint
        Dim circle As VectorDraw.Professional.vdFigures.vdCircle = New VectorDraw.Professional.vdFigures.vdCircle()
        circle.SetUnRegisterDocument(vdSC.BaseControl.ActiveDocument)
        circle.setDocumentDefaults()
        circle.Center = VectorDraw.Geometry.gPoint.MidPoint(refpoint, currentpoint)
        circle.Radius = circle.Center.Distance3D(refpoint)
        circle.Draw(act.Render)
    End Sub

    Private Sub doc_OnFigureDrawGrips(ByVal sender As Object, ByVal render As VectorDraw.Render.vdRender, ByRef cancel As Boolean) Handles doc.OnFigureDrawGrips
        If (sender Is Nothing) Then Exit Sub
        If (Not chkBox_GripDraw.Checked) Then Exit Sub
        Dim mat As VectorDraw.Geometry.Matrix = New VectorDraw.Geometry.Matrix(render.CurrentMatrix)
        render.PushToViewMatrix() ' do these 3 lines in the start of the EVENT and do not break them

        Dim fig As VectorDraw.Professional.vdPrimaries.vdFigure = sender
        Dim gripPTs As VectorDraw.Geometry.gPoints
        Dim gripPT As VectorDraw.Geometry.gPoint
        Dim circ As New VectorDraw.Professional.vdFigures.vdCircle
        circ.SetUnRegisterDocument(doc)
        circ.setDocumentDefaults()
        circ.HatchProperties = New VectorDraw.Professional.vdObjects.vdHatchProperties(VectorDraw.Professional.Constants.VdConstFill.VdFillModeSolid)
        circ.Radius = render.PixelSize * 10.0
        circ.HatchProperties.FillColor.SystemColor = Color.Red
        gripPTs = fig.GetGripPoints()
        Dim tmpPt As VectorDraw.Geometry.gPoint = New VectorDraw.Geometry.gPoint()
        For Each gripPT In gripPTs
            tmpPt = mat.Transform(gripPT)
            circ.Center = tmpPt
            circ.Draw(render)
            circ.Update()
        Next
        cancel = 1
        render.PopMatrix() ' GETTING THE VIEW MATRIX TO mat
    End Sub

    Private Sub chkBox_GripDraw_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles chkBox_GripDraw.CheckedChanged
        vdSC.BaseControl.Redraw()
    End Sub
End Class
